service/history/workflow/mutable_state_impl.go

```diff
@@ -1663,16 +1676,34 @@ func (ms *MutableStateImpl) UpdateActivity(
 
 	ms.pendingActivityInfoIDs[ai.ScheduledEventId] = ai
 	ms.updateActivityInfos[ai.ScheduledEventId] = ai
+	ms.activityInfosUserDataUpdated[ai.ScheduledEventId] = struct{}{}
 	ms.approximateSize += ai.Size() - prev.Size()
 	return nil
 }
+func (ms *MutableStateImpl) UpdateActivityTimerTaskStatus(
+	scheduleEventID int64,
+	status int32,
+) error {
+	ai, ok := ms.pendingActivityInfoIDs[scheduleEventID]
+	if !ok {
+		ms.logError(
+			fmt.Sprintf("unable to find activity event ID: %v in mutable state", scheduleEventID),
+			tag.ErrorTypeInvalidMutableStateAction,
+		)
+		return ErrMissingActivityInfo
+	}
+
+	ai.TimerTaskStatus = status
+	ms.updateActivityInfos[ai.ScheduledEventId] = ai
+	return nil
+}
 
 // UpdateActivityWithTimerHeartbeat updates an activity
 func (ms *MutableStateImpl) UpdateActivityWithTimerHeartbeat(
```

Comment:
```suggestion
func (ms *MutableStateImpl) UpdateActivityTaskStatusWithTimerHeartbeat(
```

also take activityScheduledID, timerTaskStatus and timerTimeoutVisibility as input.